### PR TITLE
Add multiline doc logic to the pretty printer

### LINKED ISSUE #1044: Pretty printer produces spurious newlines
The pretty printer inserts a spurious newline for `doc` metadata. For example,
```
❯ nickel pprint-ast <<<'{ field | doc "Hello, world" = 1 }'
{
  field | doc "
    Hello, world"
     = 1,
}
```

### PR DESCRIPTION
**Pull request by vkleen**
This PR makes the pretty printer emit multiline strings if it encounters documentation metadata containing newlines. Previously, as I complained about in #1044, the pretty printer would unconditionally insert a hard newline after the opening double quote. Additionally, it would indent the documentation string and thereby change the semantics of the code.

With this PR, multiline strings are used when necessary and single line strings are used whenever the docstring does not contain newlines. Additionally, there's a regression test using the new snapshot testing infrastructure.

Closes #1044

### CODE REVIEW

File: src/pretty.rs [hunk 1]
```diff
@@ -60,6 +60,17 @@ where
         self.text(s)
     }
 
+    fn multiline_string(&'a self, s: &str) -> DocBuilder<'a, Self, A> {
+        let percent_count = s.split(|c| c != '%').map(|v| v.len()).max().unwrap();
+        let delimiter = "%".repeat(percent_count + 1);
```
Comment (vkleen) [inline src/pretty.rs:65]: This might be too clever, but I like it 🙂

Comment (yannham) [reply src/pretty.rs:65]: FWIW, I think it's fine :slightly_smiling_face: 

It's not the best we can do (here we count any percent sign, while for example a simple modulo alone doesn't require double delimiters). A step further would be to match on sequences we know the lexer would interpret, i.e `m%"`, `"%`, and `%{`. But at least it's correct, and it's probably ok for now. Maybe add a comment to say so? (that the counts is over-approximated)

File: src/pretty.rs [hunk 2]
```diff
@@ -60,6 +60,17 @@ where
...
+        self.hardline()
+            .append(self.intersperse(
+                s.lines().map(|d| self.text(d.to_owned())),
+                self.hardline().clone(),
```
Comment (yannham) [inline src/pretty.rs:69]: Why is this clone required? It looks odd

File: src/pretty.rs [hunk 3]
```diff
@@ -60,6 +60,17 @@ where
...
+            ))
+            .enclose(format!("m{delimiter}\""), format!("\"{delimiter}"))
```
Comment (yannham) [inline src/pretty.rs:71]: I didn't even know you could do that in Rust. Has this landed recently?

File: tests/snapshot/snapshots/snapshot__pretty_multiline_doc.ncl.snap [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+    ```nickel
+      ("True" | BoolLiteral) =>
+        "true"
+      ("hello" | BoolLiteral) =>
+        error
+      (true | BoolLiteral) =>
+        error
+    ```"%%
+     = 1,
+}
```
Comment (yannham) [inline tests/snapshot/snapshots/snapshot__pretty_multiline_doc.ncl.snap:21]: Currently, we've rather written such things as:

```
{
  field | doc m%%"
    Contract to enforce the value is a string that represents a boolean literal. Additionally casts "True" to "true"
    and "False" to "false". This shouldn't interpolate: %{null}

    For example:
    ```nickel
      ("True" | BoolLiteral) =>
        "true"
      ("hello" | BoolLiteral) =>
        error
      (true | BoolLiteral) =>
        error
    ```
    "%%
     = 1,
}
```

That being said, nothing is set in stone, as we haven't decided on clear formatting rules yet.

**Review by yannham — APPROVED** [review]

File: src/pretty.rs (hunk 2 above)
Comment (vkleen) [reply src/pretty.rs:69]: Good catch, I don't think it's required. Seems to be a copy&paste artefact from the old code.

File: src/pretty.rs (hunk 3 above)
Comment (vkleen) [reply src/pretty.rs:71]: I think it's https://github.com/rust-lang/rust/pull/90473. I guess, it was stabilized in 1.58, can't really tell easily.

File: src/pretty.rs (hunk 1 above)
Comment (vkleen) [reply src/pretty.rs:65]: Actually, that was a terrible idea. We have a function `min_interpolate_sign` in the same file to compute the precise count needed 🤦

File: tests/snapshot/snapshots/snapshot__pretty_multiline_doc.ncl.snap (hunk 4 above)
Comment (vkleen) [reply tests/snapshot/snapshots/snapshot__pretty_multiline_doc.ncl.snap:21]: I've changed to your suggested formatting 👍

**Review by ebresafegaga — APPROVED** [review]

**Pull request merged**